contracts/Market.sol

```diff
@@ -1929,6 +1771,168 @@ contract Market is
+      if (removalId == 0) {
+        _incrementCurrentSupplierAddress();
+        continue;
+      }
+      uint256 removalAmount = _removal.balanceOf({
+        account: address(this),
+        id: removalId
+      });
+      if (remainingAmountToFill < removalAmount) {
+        /**
+         * The order is complete, not fully using up this removal, don't check about removing active supplier.
+         */
+        ids[countOfRemovalsAllocated] = removalId;
+        amounts[countOfRemovalsAllocated] = remainingAmountToFill;
+        suppliers[countOfRemovalsAllocated] = _currentSupplierAddress;
+        remainingAmountToFill = 0;
+      } else {
+        /**
+         * We will use up this removal while completing the order, move on to next one.
+         */
+        ids[countOfRemovalsAllocated] = removalId;
+        amounts[countOfRemovalsAllocated] = removalAmount; // this removal is getting used up
+        suppliers[countOfRemovalsAllocated] = _currentSupplierAddress;
+        remainingAmountToFill -= removalAmount;
+        /**
+        This may remove the supplier from the active suppliers and increment the current supplier address
+        * which is behavior we want in the case that the supplier is out of supply, but otherwise we do not
+        * interact with the global _currentSupplierAddress in this fulfillment flow.
+        */
+        _removeActiveRemoval({
+          removalId: removalId,
+          supplierAddress: _currentSupplierAddress
+        });
+      }
+      ++countOfRemovalsAllocated;
+      if (remainingAmountToFill == 0) {
+        break;
+      }
+    }
+    return (countOfRemovalsAllocated, ids, amounts, suppliers);
+  }
+
+  /**
+   * @notice Allocates the removals, amounts, and suppliers needed to fulfill the purchase drawing only from
+   * the specific supplier and vintages specified.
+   * @param amount The number of carbon removals to purchase.
+   * @param supplier The supplier from which to purchase carbon removals.
+   * @param vintages The vintages from which to purchase carbon removals.
+   * @return countOfRemovalsAllocated The number of distinct removal IDs used to fulfill this order.
+   * @return ids An array of the removal IDs being drawn from to fulfill this order.
+   * @return amounts An array of amounts being allocated from each corresponding removal token.
+   */
+  function _allocateSupplySingleSupplierSpecificVintages(
+    uint256 amount,
+    address supplier,
+    uint256[] memory vintages
+  )
+    private
+    returns (
+      uint256 countOfRemovalsAllocated,
+      uint256[] memory ids,
+      uint256[] memory amounts
+    )
+  {
+    RemovalsByYear storage supplierRemovalQueue = _listedSupply[supplier];
+    uint256 countOfSuppliersListedRemovals = 0;
+    for (
+      uint256 vintage = supplierRemovalQueue.earliestYear;
+      vintage <= supplierRemovalQueue.latestYear;
+      ++vintage
+    ) {
+      countOfSuppliersListedRemovals += supplierRemovalQueue
```

Comment:
Could it be optimized if we used array.fill (filled with lengths) + array.sum. 

Not sure if this is possible with our current implementation of the two array methods as this may require a callback definition

E.g. something like this  https://github.com/siromivel/lambdeth (note: don't use this exact library if you go this route as I believe I've seen an audited/optimized variant elsewhere and I'm happy to dig it up if you need it, just let me know that you're going to go that route so I can post it here as a reply)

Reply:
are you sure this would actually save gas to use an array when we don't actually need one in the first place?  I think it's more efficient to use array.sum if your data is in array in the first place and you need to sum it up, but otherwise we'll be performing unnecessary variable declarations and filling the array when really we're just adding a few numbers together and we are never going to be adding more than a small handful of numbers together (not more than the number of potential vintages supplied)

Reply:
Nope, just calling it out as a possibility (specifically because if it's a similar or equal gas cost it could enable a functional abstraction/pattern for this kind of thing in many places in our code-- and that abstraction can be optimized in assembly like the rest of arraylib). Totally fine if it doesn't make sense, but wanted to draw attention to it in case it's worth thinking about while there's relevant context for it.